src/app/clusters/resource-monitoring-server/resource-monitoring-server.cpp

```diff
@@ -257,15 +263,17 @@ CHIP_ERROR Instance::ReadReplacableProductList(AttributeValueEncoder & aEncoder)
         productListManagerInstance->Reset();
 
         err = aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
-            Attributes::ReplacementProductStruct::Type replacementProductStruct;
-            while (productListManagerInstance->Next(replacementProductStruct) == CHIP_NO_ERROR)
+            Attributes::GenericReplacementProductStruct::GenericType replacementProductStruct;
+            CHIP_ERROR iteratorError = CHIP_NO_ERROR;
+            while (CHIP_NO_ERROR == iteratorError)
             {
                 ReturnErrorOnFailure(encoder.Encode(replacementProductStruct));
```

Comment:
Has this been tested?  This does not look like it does the right thing.  Please test this code.

Reply:
Tested manually, had missed the bug. Added a test for this as well